### PR TITLE
feat: docker image builds

### PR DESCRIPTION
**Pull request by nugaon**

### CODE REVIEW

**Comment by agazso** [comment]
There are vulnerabilities reported by `npm` when running `install`. Not sure if they are indeed vulnerabilities but it would be good to update the dependencies.

```
% npm i

added 1783 packages, and audited 1784 packages in 2m

28 vulnerabilities (19 low, 8 moderate, 1 high)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
```

**Comment by agazso** [comment]
I suggest to use the [ShellCheck](https://www.shellcheck.net/) tool. It is like a linter for shell script and it also has a VS Code [extension](https://marketplace.visualstudio.com/items?itemName=timonwong.shellcheck).

The most common suggestion it provides is to use `$(...)` instead of backticks: https://github.com/koalaman/shellcheck/wiki/SC2006

It also helps finding unquoted variables and other common issues.

File: scripts/utils/env-variable-value.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+echo_env_variable_value() {
```
Comment (agazso) [inline scripts/utils/env-variable-value.sh:3]: This seems to me a bit fragile as an alternative to sourcing the `.env` file. When sourcing files comments are allowed but this script does not handle comments.

File: scripts/utils/env-variable-value.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  local -n REF=$1
```
Comment (agazso) [inline scripts/utils/env-variable-value.sh:4]: On macOS the default bash version installed is 3.2 which does not support the `-n` option to the `local` command.

Consider using this instead: `local REF=$(printenv $1)`

**Comment by agazso** [comment]
After following the setup with the default config without errors, I get an error when trying to start the environment and there are no peers connected to my node:

```
% ./scripts/network.sh
% ./scripts/blockchain.sh 
<output omitted>
% npm run migrate:chequebook
<output omitted>
% npm run supply
<output omitted>
% ./scripts/bee-docker-build.sh
<output omitted>
% ./scripts/blockchain-docker-build.sh
<output omitted>
% ./scripts/environment.sh start
Start Blockchain node...
docker: Error response from daemon: Conflict. The container name "/swarm-test-blockchain" is already in use by container "c582496d5379692d00659bb9c15092f61badb91897e3f908e0ba7236780c9a1a". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
Start Bee nodes...
start Bee Queen process
b3867acbc9e6afbc077b72035c3df473ebdebe1690ab4b6acfb07c950bfb23a7
start Bee worker 1 process
a3ff5768c531ee7ac407c63fe5eb08c980e05527dc4441637aae50192ed921ee
start Bee worker 2 process
28333fb6aebd68bf7ab5cfc0b97681cff5cdc16fdc20b3cd92e669715703b28b
start Bee worker 3 process
a97afbea9f23d4056fddab8c576496cff54d889fab51b5fb16786b11c215a9a0
start Bee worker 4 process
8246de6939f9276122b85a2f60144d8b3edb69bc5511bfcf028536de81d4c288
  (     /   /  )%%%%%%%(  \   \     )
  (___/___/__/           \__\___\___)
     (     /  /(%%%%%%%)\  \     )
      (__/___/ (%%%%%%%) \___\__)
              /(       )\
            /   (%%%%%)   \
                 (%%%)
                   !                   
time="2021-04-28T10:31:38Z" level=info msg="version: 0.5.3-acbd0e2"
time="2021-04-28T10:31:38Z" level=warning msg="clef is not enabled; portability and security of your keys is sub optimal"
time="2021-04-28T10:31:38Z" level=info msg="using existing swarm network address: 18d0987d8d012cd3ca589d2df13fa28eeb8961f543faeb5a991df8b6c4c25333"
time="2021-04-28T10:31:38Z" level=info msg="swarm public key 03d5f6051105af2bc1dfb657031fc9d10df9be44ed24608e730a9a0cd07c97e906"
time="2021-04-28T10:31:39Z" level=debug msg="using existing libp2p key"
time="2021-04-28T10:31:39Z" level=debug msg="using existing pss key"
time="2021-04-28T10:31:39Z" level=info msg="pss public key 02f20bfb14a31491f8a9aec2dc455be4353733040e921ececcf612aca735b0bc7c"
time="2021-04-28T10:31:39Z" level=info msg="using ethereum address 26234a2ad3ba8b398a762f279b792cfacd536a3f"
time="2021-04-28T10:31:39Z" level=info msg="debug api address: [::]:1635"
time="2021-04-28T10:31:39Z" level=info msg="using custom factory address: 5b1869d9a4c187f2eaa108f3062412ecf0526b24"
time="2021-04-28T10:31:39Z" level=info msg="no chequebook found, deploying new one."
time="2021-04-28T10:31:39Z" level=info msg="deploying new chequebook in transaction 30a0f03020dc1a0d89b126902fc51ec010e668ab4d696a5d399bec1da015b3a1"
time="2021-04-28T10:31:40Z" level=info msg="deployed chequebook at address 3c63250afa2470359482d98749f2d60d2971c818"
time="2021-04-28T10:31:40Z" level=info msg="depositing 100000000000000000 token into new chequebook"
time="2021-04-28T10:31:41Z" level=info msg="sent deposit transaction d8f8887f592a657968501484f9af8fb1497ef88026b3309a6c295a6d1827d8e8"
time="2021-04-28T10:31:42Z" level=info msg="successfully deposited to chequebook"
time="2021-04-28T10:31:44Z" level=debug msg="initializing NAT manager"
time="2021-04-28T10:31:50Z" level=debug msg="NAT manager initialized"
time="2021-04-28T10:31:50Z" level=debug msg="p2p address: /ip4/172.19.0.3/tcp/1634/p2p/16Uiu2HAmBTraFFjULgNxfUm4DbCjuejN8KFhx9xAdE4kKRt82ccL"
time="2021-04-28T10:31:50Z" level=debug msg="p2p address: /ip4/127.0.0.1/tcp/1634/p2p/16Uiu2HAmBTraFFjULgNxfUm4DbCjuejN8KFhx9xAdE4kKRt82ccL"
time="2021-04-28T10:31:50Z" level=info msg="database capacity: 5000000 chunks (approximately 19.5GB)"
time="2021-04-28T10:31:50Z" level=info msg="name resolver: no name resolution service provided"
time="2021-04-28T10:31:50Z" level=info msg="api address: [::]:1633"
time="2021-04-28T10:31:50Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:32:20Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:32:36Z" level=info msg="api access" duration=9.3446e-05 ip=172.19.0.1 method=GET proto=HTTP/1.1 size=19 status=200 uri=/ user-agent=curl/7.54.0
time="2021-04-28T10:32:39Z" level=info msg="debug api access" duration=0.000257777 ip=172.19.0.1 method=GET proto=HTTP/1.1 size=36 status=404 uri=/ user-agent=curl/7.54.0
time="2021-04-28T10:32:50Z" level=info msg="debug api access" duration=0.000177197 ip=172.19.0.1 method=GET proto=HTTP/1.1 size=14 status=200 uri=/peers user-agent=curl/7.54.0
time="2021-04-28T10:32:50Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:33:20Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:33:50Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:34:20Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:34:50Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:35:20Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:35:50Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:36:20Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:36:50Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
time="2021-04-28T10:37:20Z" level=debug msg="kademlia has no connected peers, trying bootnodes"
```

**Comment by nugaon** [comment]
> I suggest to use the [ShellCheck](https://www.shellcheck.net/) tool. It is like a linter for shell script and it also has a VS Code [extension](https://marketplace.visualstudio.com/items?itemName=timonwong.shellcheck).
> 
> The most common suggestion it provides is to use `$(...)` instead of backticks: https://github.com/koalaman/shellcheck/wiki/SC2006
> 
> It also helps finding unquoted variables and other common issues.

Unfortunately all of them comes from `truffle` they have a fresh issue about it: https://github.com/trufflesuite/truffle/issues/3986

**Comment by nugaon** [comment]
> After following the setup with the default config without errors, I get an error when trying to start the environment and there are no peers connected to my node:

Try out with fresh local docker images regarding to the env, without any running/stopped container that you used in the previous runs. Issue could be after reading your logs:

> docker: Error response from daemon: Conflict. The container name "/swarm-test-blockchain" is already in use by container "c582496d5379692d00659bb9c15092f61badb91897e3f908e0ba7236780c9a1a". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.

Also don't forget to wipe out the `localstore` and `statestore` folders from the [scripts/bee-data-dirs](scripts/bee-data-dirs) folder when you make the docker images. for these, you can use the `bee-cleanup.sh` script.

File: scripts/utils/env-variable-value.sh (hunk 1 above)
Comment (nugaon) [reply scripts/utils/env-variable-value.sh:3]: I prefer the global env variables over the .env file's variables, that's why I wrote this script.
Nevertheless your comment is justified, I make some adjustions

**Comment by agazso** [comment]
> Try out with fresh local docker images regarding to the env, without any running/stopped container that you used in the previous runs. Issue could be after reading your logs:

I tried again after running the cleanup script and pruned the docker containers. There are no errors now, but there are still no connected peers. Any idea how can I debug to see what's wrong?

File: scripts/utils/env-variable-value.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+echo_env_variable_value() {
+  REF=$(printenv $1)
+
+  # There were no global set
+  if [ -z "$REF" ] ; then
+    MY_PATH=$(dirname "$0")
+    ENV_PATH=$( cd "$MY_PATH/.." && pwd )
+    VALUE=$(cat "$ENV_PATH/.env" | grep "^$1" | cut -d= -f2)
```
Comment (agazso) [inline scripts/utils/env-variable-value.sh:10]: `cat $filename | grep $pattern` can be rewritten to `grep $pattern $filename`

**Comment by agazso** [comment]
ShellCheck reports a lot of unquoted variables. See the explanation here:
https://github.com/koalaman/shellcheck/wiki/SC2086

I think it's a good idea to enforce this.

**Comment by agazso** [comment]
Another thing to consider to use `-e` flag to bash so that it terminates on errors. It can be added to the hashbang command at the top of the files:
```
#!/bin/bash -e
```
The rationale is that errors currently are ignored and the scripts continue execution which can lead to unexpected situations.

**Review by agazso — APPROVED** [review]

**Pull request merged**